bundler/lib/bundler/lockfile_parser.rb

```diff
@@ -224,7 +224,7 @@ def parse_spec(line)
 
         # Avoid introducing multiple copies of the same spec (caused by
         # duplicate GIT sections)
-        @specs[@current_spec.identifier] ||= @current_spec
+        @specs[@current_spec.hash] ||= @current_spec
```

Comment:
Wouldn't this hash cause the same issues we saw with the `DepProxy` hash? Should this maybe use the constituents of the hash? (don't know how much this code is called, and I'm guessing the previous implementation has worked relatively well to date, but the risk remains).

Reply:
Hei @lukaso, thanks for your comment.

I'm really not sure. Thinking about this issue more, my understanding is that as long as you rely on `#hash` values being generated by ruby, then it is irrelevant whether you use `#hash` or the raw object instance as the key to a hash, since ruby will always use its internally generated hash vale that it's guaranteed to be unique. If you implement your own `#hash` method though, you need to make sure that the result cannot have collissions. However, [the docs on user-defined #hash values](https://ruby-doc.org/core-3.0.0/Hash.html#class-Hash-label-User-Defined+Hash+Keys) make it seem like the implementation I'm using here (XOR'ing the `#hash` value of the different data components of the object) is equally safe, so I tend to say this is fine.

That said, I'm fine with reverting this change and restoring the `Identifier` class too.

In any case, my plan is to move these changes to #4277 and leave on the dep proxy comparison fix in this PR. I'll do this on Monday.

Reply:
From what I can tell, your implementation is exactly as good  as the `Identifier` class one you replaced in terms of `hash`. That was also using hash. So no need to revert. Both implementations have the same potential flaw. I think you'd have to use an array of the actual values as the key to `@specs` in order to avoid the issue, exactly as we are doing on `DepProxy`. And then there's the question of how often this code is called (could it cause a performance issue?).

Reply:
> From what I can tell, your implementation is exactly as good as the Identifier class one you replaced in terms of hash. That was also using hash. 

Was it? I don't see a custom `#hash` in the class.

Anyways, now I understand that your comment is the exact same concern raised in https://github.com/rubygems/rubygems/pull/4216#discussion_r553545164. Just like in that discussion, I don't believe there's an issue here, but I will change the code as per your request. It's going to look a bit weird though. It'd be

```suggestion
        @specs[[@current_spec.name, @current_spec.version, @current_spec.platform_string] ||= @current_spec
```

Correct?

Regarding the performance, I don't understand how this could cause a performance issue, even if it was a hot path. I believe it's not since this is the lockfile parser, and lockfile parsing should happen only once, but still.

Reply:
Yes, correct, that was the concern (not raised by me, but noticed same behaviour here).

Reply:
Great, I will update this as we did there. For now I removed all these changes from this PR to focus on the dependency requirement fix. I'll move the rest of the improvements to #4277 after merging this one (I'll merge it after CI).

Reply:
Just to complete my further thoughts on this issue, it seems that ruby properly resolves collissions as long as you provide a proper `eql?` definition. So, in the example given at https://ruby-doc.org/core-3.0.0/Hash.html#class-Hash-label-User-Defined+Hash+Keys:

```ruby
class Book
  attr_reader :author, :title

  def initialize(author, title)
    @author = author
    @title = title
  end

  def ==(other)
    self.class === other &&
      other.author == @author &&
      other.title == @title
  end

  alias eql? ==

  def hash
    @author.hash ^ @title.hash # XOR
  end
end

book1 = Book.new 'matz', 'Ruby in a Nutshell'
book2 = Book.new 'matz', 'Ruby in a Nutshell'

reviews = {}

reviews[book1] = 'Great reference!'
reviews[book2] = 'Nice and compact!'

reviews.length #=> 1
```

You can check forcing a `#hash` collision and still getting the expected behaviour:

```ruby
book3 =  Book.new 'Ruby in a Nutshell', 'matz'

book3.hash == book2.hash # => true

reviews[book3] = "Different"

reviews.lenght # => 2
```

So, again, no issue here.

Reply:
@lukaso I moved all these changes to #4277. This should be fixed there!